File: Code/Tests/RerF_test.m
% train and test clas

clear
close all
clc

nTrials = 20;

FilePath = '/scratch/groups/jvogels3/tyler/Data/tests/';
OutFile = '/scratch/groups/jvogels3/tyler/RandomerForest/Results/Tests/test_results_2017_08_01.mat';

Datasets = {'Sparse_parity';'Trunk';'Orthant';'mnist'};

for d = 1:length(Datasets)
    D = Datasets{d};

    TrainFile = [FilePath D '_train.csv'];
    TestFile = [FilePath D '_test.csv'];

    Xtrain = dlmread(TrainFile);
    Ytrain = cellstr(num2str(Xtrain(:,end)));
    Xtrain(:,end) = [];

    Xtest = dlmread(TestFile);
    Ytest = cellstr(num2str(Xtest(:,end)));
    Xtest(:,end) = [];

    [ntrain,p] = size(Xtrain);
    ntest = length(Ytest);

    Labels = unique([Ytrain;Ytest]);
    nClasses = length(Labels);

    if strcmp(D, 'mnist')
        mtrys = ceil([p^(1/2) 10*p]);
    else
        mtrys = ceil(p.^[1/2 2]);
    end

    Params.(D).nTrees = 500;
    Params.(D).Stratified = true;
    Params.(D).NWorkers = 16;
    Params.(D).Rescale = 'off';
    Params.(D).ForestMethod = 'rerf';
    Params.(D).RandomMatrix = 'binary-redundant';
    Params.(D).d = mtrys;
    Params.(D).rho = 1/p;
    Params.(D).Rotate = false;

    Train_Time = NaN(nTrials,length(Params.(D).d)*length(Params.(D).rho));
    OOB_Error = NaN(nTrials,length(Params.(D).d)*length(Params.(D).rho));
    Num_Nodes = NaN(nTrials,length(Params.(D).d)*length(Params.(D).rho));
    Test_Error = NaN(nTrials,length(Params.(D).d)*length(Params.(D).rho));

    for trial = 1:nTrials

        % train classifier
        poolobj = gcp('nocreate');
        if isempty(poolobj)
            parpool('local',Params.(D).NWorkers,...
                'IdleTimeout',360);
        end

        [Forest,~,Train_Time(trial,:)] = ...
            RerF_train(Xtrain,Ytrain,Params.(D));

        fprintf('Training complete\n')

        % evaluate classifier

        for k = 1:length(Forest)
            Labels = Forest{k}.classname;
            nClasses = length(Labels);
            Scores = rerf_oob_classprob(Forest{k},Xtrain,'last');
            Predictions = predict_class(Scores,Labels);
            OOB_Error(trial,k) = misclassification_rate(Predictions,Ytrain,false);
            NN = NaN(Forest{k}.nTrees,1);
            Trees = Forest{k}.Tree;
            parfor kk = 1:Forest{k}.nTrees
                NN(kk) = Trees{kk}.numnodes;
            end
            Num_Nodes(trial,k) = mean(NN);

            if ~strcmp(Forest{k}.Rescale,'off')
                TestScores = rerf_classprob(Forest{k},Xtest,'last',Xtrain);
            else
                TestScores = rerf_classprob(Forest{k},Xtest,'last');
            end
            TestPredictions = predict_class(TestScores,Forest{k}.classname);
            Test_Error(trial,k) = misclassification_rate(TestPredictions,Ytest,false);
        end

        clear Forest
    end

    TrainTime.(D) = Train_Time;
    OOBError.(D) = OOB_Error;
    TestError.(D) = Test_Error;
    NumNodes.(D) = Num_Nodes;
end

save(OutFile,'Params','TrainTime','OOBError','TestError','NumNodes')

delete(gcp);

File: R/Code/Tests/RerF_test.R
rm(list = ls())
source("~/work/tyler/RerF/Code/Classifiers/rfr_function.R")

nTrials <- 20L

filePath <- "~/work/tyler/Data/tests/"
outFile <- "~/work/tyler/RerF/Results/Tests/test_results_2017_08_22.RData"
# filePath <- "~/tests/"
# outFile <- "~/RerF/Results/Tests/test_results_2017_08_22.RData"

dataSets <- c("Sparse_parity", "Trunk", "Orthant", "mnist")

trainTime <- list()
OOBError <- list()
numNodes <- list()
testError <- list()
seed <- list()

for (ds in 1:length(dataSets)) {
# for (ds in 4L) {
  D <- dataSets[ds]
  print(paste("Dataset:", D))
  
  trainFile <- paste(filePath, D, "_train.csv", sep = "")
  testFile <- paste(filePath, D, "_test.csv", sep = "")
  
  Xtrain <- as.matrix(read.table(trainFile, header = F, sep = ",", quote = "", row.names = NULL))
  ntrain <- nrow(Xtrain)
  p <- ncol(Xtrain) - 1L
  Ytrain <- as.integer(Xtrain[, p + 1L]) + 1L
  Xtrain <- Xtrain[, -(p + 1L)]
  
  Xtest <- as.matrix(read.table(testFile, header = F, sep = ",", quote = "", row.names = NULL))
  ntest <- nrow(Xtest)
  Ytest <- as.integer(Xtest[, p + 1L] + 1L)
  Xtest <- Xtest[, -(p + 1L)] 
  
  labels <- unique(c(Ytrain,Ytest))
  nClasses <- length(labels)
  
  if (D == "mnist") {
    mtrys <- ceiling(c(p^(1/2), 10*p))
  } else {
    mtrys <- ceiling(p^c(1/2, 2))
  }
  # mtrys <-p^2
  # mtrys <- p
  # mtrys <- 10*p
  # mtrys <- ceiling(sqrt(p))
    
  nTrees <- 500L
  stratify <- T
  replacement <- T
  randomMatrix <- "binary"
  sparsity <- 1/p
  rotate <- F
  MinParent <- 2L
  MaxDepth <- "inf"
  COOB <- T
  CNS <- F
  NumCores <- 16L
  comp.mode <- "batch"
  rank.transform <- F
  
  train.time <- matrix(0, nTrials, length(mtrys))
  oob.error <- matrix(0, nTrials, length(mtrys))
  num.nodes <- matrix(0L, nTrials, length(mtrys))
  test.error <- matrix(0, nTrials, length(mtrys))
  seed[[D]] <- matrix(sample.int(1e4L, nTrials*length(mtrys)), nTrials, length(mtrys))
  
  for (trial in 1:nTrials) {
    print(paste("Trial", trial))
    for (m in 1:length(mtrys)) {
      # train the classifier
      cat("training\n")
      start.time <- proc.time()
      forest <- rerf(Xtrain, Ytrain, MinParent = MinParent, trees = nTrees, MaxDepth = MaxDepth, replacement = replacement,
                     stratify = stratify, FUN = randmat, options = list(p, mtrys[m], randomMatrix, sparsity), COOB = COOB,
                     CNS = CNS, NumCores = NumCores, seed = seed[[D]][trial, m], rotate = rotate, rank.transform = rank.transform)
      train.time[trial, m] <- (proc.time() - start.time)[[3]]
      cat("training complete\n")
      cat(paste0("elapsed time: ", train.time[trial, m], " sec\n"))
      
      # compute out-of-bag predictions
      cat("OOB\n")
      scores <- OOBpredict(Xtrain, forest, NumCores, comp.mode = comp.mode)
      oob.error[trial, m] <- mean(max.col(scores) != Ytrain)
      
      # compute average number of nodes per tree
      cat("node size\n")
      num.nodes[trial, m] <- mean(sapply(forest, FUN = function(tree) length(tree$treeMap)))
      
      # compute predictions on test set  
      cat("test predictions\n")
      scores <- predict(Xtest, forest, NumCores, comp.mode = comp.mode)
      test.error[trial, m] <- mean(max.col(scores) != Ytest)
      rm(forest)
      gc()
    }
  }
  trainTime[[D]] <- train.time
  OOBError[[D]] <- oob.error
  numNodes[[D]] <- num.nodes
  testError[[D]] <- test.error
}

save(file = outFile, list = c("trainTime", "OOBError", "numNodes", "testError", "seed"))
